File: Helper_Scripts/Awtrix_Clone.py
#Clones the ICONS and MELODIES folder from one to another AWTRIX Light

import os
import argparse
import requests

def copy_files(source_ip, target_ip, directory):
    print(f'Copying files from directory {directory}')

    # Retrieve file list from source server
    response = requests.get(f'http://{source_ip}/list?dir=/{directory}')
    file_list = response.json()

    # Iterate through the file list
    for file_info in file_list:
        if file_info['type'] == 'file':
            filename = file_info['name']
            source_file_url = f'http://{source_ip}/{directory}/{filename}'
            target_filename = f'/{directory}/{filename}'
            target_file_url = f'http://{target_ip}/edit?filename={target_filename}'

            # Download the file
            response = requests.get(source_file_url)
            if response.status_code == 200:
                # Upload the file to the target server
                files = {'file': (target_filename, response.content)}
                response = requests.post(target_file_url, files=files)
                print(f'{filename}: {response.text}')
            else:
                print(f'Failed to download file {filename}')

# Create argument parser
parser = argparse.ArgumentParser(description="Copy files from one server to another.")
parser.add_argument("source_ip", help="IP address of the source server.")
parser.add_argument("target_ip", help="IP address of the target server.")

# Parse arguments
args = parser.parse_args()

# Copy files from the directories /ICONS and /MELODIES
copy_files(args.source_ip, args.target_ip, 'ICONS')
copy_files(args.source_ip, args.target_ip, 'MELODIES')

File: Helper_Scripts/Awtrix_Clone.go
package main

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"bytes"
	"encoding/json"
)

func copyFiles(sourceIP, targetIP, directory string) {
	fmt.Printf("Copying files from directory %s\n", directory)

	// Retrieve file list from source server
	response, err := http.Get(fmt.Sprintf("http://%s/list?dir=/%s", sourceIP, directory))
	if err != nil {
		fmt.Printf("Failed to retrieve file list from source server: %s\n", err.Error())
		return
	}
	defer response.Body.Close()

	// Read the response body
	body, err := ioutil.ReadAll(response.Body)
	if err != nil {
		fmt.Printf("Failed to read response body: %s\n", err.Error())
		return
	}

	// Parse the JSON response
	fileList := make([]map[string]interface{}, 0)
	err = json.Unmarshal(body, &fileList)
	if err != nil {
		fmt.Printf("Failed to parse JSON response: %s\n", err.Error())
		return
	}

	// Iterate through the file list
	for _, fileInfo := range fileList {
		fileType, ok := fileInfo["type"].(string)
		if !ok {
			fmt.Println("Failed to get file type")
			continue
		}

		if fileType == "file" {
			filename, ok := fileInfo["name"].(string)
			if !ok {
				fmt.Println("Failed to get filename")
				continue
			}

			sourceFileURL := fmt.Sprintf("http://%s/%s/%s", sourceIP, directory, filename)
			targetFilename := fmt.Sprintf("/%s/%s", directory, filename)
			targetFileURL := fmt.Sprintf("http://%s/edit?filename=%s", targetIP, targetFilename)

			// Download the file
			response, err := http.Get(sourceFileURL)
			if err != nil {
				fmt.Printf("Failed to download file %s: %s\n", filename, err.Error())
				continue
			}
			defer response.Body.Close()

			if response.StatusCode == http.StatusOK {
				// Read the file content
				fileContent, err := ioutil.ReadAll(response.Body)
				if err != nil {
					fmt.Printf("Failed to read file content: %s\n", err.Error())
					continue
				}

				// Upload the file to the target server
				response, err := http.Post(targetFileURL, "application/octet-stream", bytes.NewBuffer(fileContent))
				if err != nil {
					fmt.Printf("Failed to upload file %s: %s\n", filename, err.Error())
					continue
				}

				fmt.Printf("%s: %s\n", filename, response.Status)
			} else {
				fmt.Printf("Failed to download file %s: %s\n", filename, response.Status)
			}
		}
	}
}

func main() {
	// Parse command-line arguments
	if len(os.Args) != 3 {
		fmt.Println("Usage: awtrix_clone source_ip target_ip")
		return
	}

	sourceIP := os.Args[1]
	targetIP := os.Args[2]

	copyFiles(sourceIP, targetIP, "ICONS")
	copyFiles(sourceIP, targetIP, "MELODIES")
}
